### ISSUE TITLE
`yield` inside `with` blocks with type comment: Type does not follow the type comment

### ISSUE THREAD

**Issue by xuhdev** [python/mypy#9869]
**Bug Report**

Consider the situation where, inside a function, a type comment is provided for a `with` statement and a `yield` statement is in the `with` block. When comparing the yield type and the type annotation of the function, mypy ignores the type comment on the `with` line.

**To Reproduce**

Save the following code as `a.py` and run `mypy a.py`.

```python
from __future__ import annotations
from typing import Iterator

class Context:
    def __enter__(self) -> Context:
        return self

    def __exit__(self, *args):
        pass


class ContextChild(Context):
    pass


def call_context() -> Iterator[ContextChild]:
    with ContextChild() as m:  # type: ContextChild
        yield m
        reveal_locals()
```

**Expected Behavior**

No issue should be reported.

**Actual Behavior**

The following is reported:

```
a.py:17: error: Incompatible types in assignment (expression has type "Context", variable has type "ContextChild")
a.py:19: note: Revealed local types are:
a.py:19: note:     m: a.ContextChild
Found 1 error in 1 file (checked 1 source file)
```

Mypy correctly recognizes `m` as a `ContextChild` object, but when it comes to compare the yield type and the annotation of `call_context`, it ignores the type comment.

**Your Environment**

- Mypy version used: 0.790
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: Python 3.7.3, 3.9.0
- Operating system and version: Debian 10

**Comment by hauntsaninja** [comment]
mypy's correct here. You might want to use a self type for the return of `__enter__`.

**Comment by xuhdev** [comment]
@hauntsaninja why is it correct? The comment `type: ContextChild` is ignored in the reported error.

**Comment by hauntsaninja** [comment]
For the same reason that `m: ContextChild = Context()` would complain. For more questions, try Gitter
